### ISSUE TITLE
Discrepancy in BCEWithLogitsLoss and ClassNLLLoss

### ISSUE THREAD

**Issue by shubhamjain0594** [pytorch/pytorch#2180]
In BCEWithLogitsLoss, weights are unnormalized and used, while in ClassNLLLoss weights are normalized and used. 

Some discussion around this can be found [here](https://github.com/pytorch/pytorch/commit/67968cb60b1d3021834594967d4140a36a8213e3#commitcomment-23155134). Is this discrepancy expected? Else if this is an issue I can send a PR to resolve this.

**Comment by soumith** [comment]
this is not an expected discrepancy. Maybe BCEWithLogitsLoss can be fixed.

**Comment by Kaixhin** [comment]
@apaszke Your ENet paper used a class weighting scheme that was not normalised if I remember correctly - were they normalised during the  implementation or did you keep them explicitly as-is?

**Comment by varunagrawal** [comment]
@Kaixhin the weights were normalized. [Here's](https://github.com/e-lab/ENet-training/blob/f76214475ace6df38d50fa93a334ed7ffcc93af6/train/models/encoder.lua#L116) the code snippet.

**Comment by zou3519** [comment]
@shubhamjain0594 I think comparing BCEWithLogitsLoss (and BCELoss) weights with NLLLoss weights isn't a fair comparision: NLLLoss weights are per-class while BCELoss weights are per batch-element. 

In this sense I don't think it makes sense to normalize the weights in BCEWithLogitsLoss because they're per-batch element and the use-case for them is weighing the elements in the loss differently: what do you think?
